website/project/model.py

```diff
@@ -3621,81 +3681,97 @@ class DraftRegistration(AddonModelMixin, StoredObject):
         # if self.is_pending_review or self.is_approved:
         #    raise NodeStateError('Cannot edit while this draft is being reviewed')
+
         changes = []
-        for key, value in metadata.iteritems():
-            old_value = self.registration_metadata.get(key)
+        for question_id, value in metadata.iteritems():
+
+            old_value = self.registration_metadata.get(question_id)
+
+            if old_value:
+                old_comments = old_value.get('comments', [])
+                new_comments = value.get('comments', [])
+
+                #  we are using the `created` attribute sort of as a primary key
+                old_comment_ids = [comment['created'] for comment in old_comments]
+
+                # Handle comment conflicts
+                for old_comment in old_comments:
+                    for new_comment in new_comments:
+                        new_id = new_comment.get('created', [])
+                        # if the primary key is already in use and the old comment is more recent,
+                        if new_id in old_comment_ids and old_comment['lastModified'] > new_comment['lastModified']:
+                            # use the old one instead
+                            loc = new_comments.index(new_comment)
+                            new_comments[loc] = old_comment
+
             if not old_value or old_value.get('value') != value.get('value'):
-                changes.append(key)
+                changes.append(question_id)
+
         self.registration_metadata.update(metadata)
+
         if save:
             self.save()
-        # TODO: uncomment to nullify approval state if edited
-        # project_signals.draft_edited.send(self, changes)
-        # self.after_edit(changes)
+        project_signals.draft_edited.send(changes)
+        self.after_edit(changes)
         return changes
 
     def before_edit(self, auth):
+        # TODO(samchrisinger): Make sure we still need this
         messages = []
-        # TODO: uncomment to provide pre-edit warnings for drafts that are approved or are pending approval
-        # if self.flags.get('isApproved'):
-        #     messages.append('The draft registration you are editing is currently approved. Please note that if you make any changes (excluding comments) this approval status will be revoked and you will need to submit for approval again.')
-        # if self.flags.get('isPendingReview'):
-        #     messages.append('The draft registration you are editing is currently pending review. Please note that if you make any changes (excluding comments) this request will be cancelled and you will need to submit for approval again.')
+        if self.is_approved:
+            messages.append('The draft registration you are editing is currently approved. Please note that if you make any changes (excluding comments) this approval status will be revoked and you will need to submit for approval again.')
+        if self.flags.get('isPendingReview'):
```

Comment:
Is this flag expected to be present on all forms of draft registrations? If so it might be a good idea to make it an explicit field on the model rather than part of a less-structured dictionary field.